crates/store/README.md

```diff
@@ -0,0 +1,829 @@
+  - [Detailed component descriptions](#detailed-component-descriptions)
+    - [Database interface](#database-interface-1)
+    - [Key structure](#key-structure-1)
+    - [Data types and entries](#data-types-and-entries-1)
+    - [Layers](#layers-1)
+    - [Transactions](#transactions-1)
+    - [Iterators](#iterators-1)
+
+## Introduction
+
+The state/store library is a versatile and efficient key-value storage system
+designed for managing application state and transactions. It provides a flexible
+and extensible architecture for storing, retrieving, and manipulating data
+across different storage backends.
+
+In this context, "state" refers to the data being stored and managed, while
+"store" refers to the actual storage mechanism and its associated operations.
+The library abstracts these concepts into a unified interface, allowing
+developers to work with application state seamlessly. From here on, the term
+"store library" or just "store" will be used, but this encompasses state too.
+
+Key features of the store library include:
+
+  - Support for potentially multiple storage backends (currently just
+    [RocksDB][])
+  - Transactional operations with atomic commits
+  - Flexible key-value structure with custom data types
+  - Iteration and range queries
+  - Temporal layers for managing changes before committing
+
+## Core components
+
+### Database interface
+
+The `Database` trait defines the core interface for interacting with the
+underlying storage system. It provides methods for basic CRUD operations,
+iteration, and transaction application.
+
+Key operations:
+
+  - `open()`: Initialise the database with a given configuration
+  - `has()`, `get()`, `put()`, `delete()`: Basic key-value operations
+  - `iter()`: Create an iterator for range queries
+  - `apply()`: Apply a transaction to the database
+
+The `RocksDB` struct implements this trait, providing a concrete implementation
+using the [RocksDB][] storage engine.
+
+### Key structure
+
+The library uses a flexible key structure defined by the `Key` type and
+associated traits:
+
+  - `KeyComponent`: Defines a part of a key with a specific length
+  - `KeyComponents`: Combines multiple `KeyComponent`s into a complete key
+  - `AsKeyParts`: Converts a key into its column and components
+  - `FromKeyParts`: Constructs a key from its components
+
+This structure allows for efficient encoding and decoding of complex keys,
+supporting various data models.
+
+### Data types and entries
+
+The `DataType` trait defines how values are serialised and deserialised. The
+`Entry` trait combines a key with its associated data type, providing a complete
+representation of a stored item.
+
+Predefined entry types include:
+
+  - `ContextMeta`: Metadata for a context
+  - `ContextState`: State data for a context
+  - `ContextIdentity`: Identity information for a context
+  - `ContextTransaction`: Transaction data for a context
+  - `Generic`: A generic key-value pair
```

Comment:
I noticed you removed `Generic` here, to clarify.. my comment wasn't about this, but mentioning predefined entries at all at this level

And that's only in question because when you discuss keys a section above, we didn't talk about the types of keys.

Figured it's worth clarifying just in case it wasn't clear, it's not that I think it's wrong to mention or there's something wrong with `Generic` specifically.

Not sure what model is best, since (due to `Predefined`'s behaviour `Keys` are treated as entries, so it might seem like duplication.

Unless there's some way of talking about both cases.

Reply:
Hmmmm, I see them as different... I am not sure what outcome is best.

For now I've re-added `Generic` - thanks for clarifying that.